presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ShardJoins.java

```diff
@@ -0,0 +1,212 @@
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final StatsCalculator statsCalculator;
+    private boolean isEnabledForTesting;
+
+    public ShardJoins(Metadata metadata, FunctionAndTypeManager functionAndTypeManager, StatsCalculator statsCalculator)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+        this.statsCalculator = requireNonNull(statsCalculator, "statsCalculator is null");
+    }
+
+    @Override
+    public void setEnabledForTesting(boolean isSet)
+    {
+        isEnabledForTesting = isSet;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isEnabledForTesting || !getShardedJoinStrategy(session).equals(DISABLED);
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (isEnabled(session)) {
+            return SimplePlanRewriter.rewriteWith(new Rewriter(session, metadata, functionAndTypeManager, idAllocator, variableAllocator), plan, new HashSet<>());
+        }
+
+        return plan;
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Set<VariableReferenceExpression>>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final PlanNodeIdAllocator planNodeIdAllocator;
+        private final VariableAllocator planVariableAllocator;
+        private Rewriter(Session session, Metadata metadata,
+                FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator planNodeIdAllocator, VariableAllocator planVariableAllocator)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.planNodeIdAllocator = requireNonNull(planNodeIdAllocator, "planNodeIdAllocator is null");
+            this.planVariableAllocator = requireNonNull(planVariableAllocator, "planVariableAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode joinNode, RewriteContext<Set<VariableReferenceExpression>> context)
+        {
+            if (isApplicable(joinNode)) {
+                long numShards = getNumberOfShards();
+                RowExpression randomNumber = call(
+                        functionAndTypeManager,
+                        "random",
+                        BIGINT,
+                        constant((long) numShards - 1, BIGINT));
```

Comment:
Looks like that the argument here is exclusive, i.e.  numShards - 1 will not be a possible output. Can you double check and change if needed?

Reply:
thanks for spotting, will fix!
It wasn't incorrect but might be inefficient as we were generating less shards on the left side of the join